spark3/src/main/java/org/apache/iceberg/actions/Spark3MigrateAction.java

```diff
@@ -0,0 +1,144 @@
+import org.apache.iceberg.spark.SparkSessionCatalog;
+import org.apache.iceberg.spark.SparkTableUtil;
+import org.apache.iceberg.spark.source.SparkTable;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.catalyst.TableIdentifier;
+import org.apache.spark.sql.catalyst.analysis.NoSuchTableException;
+import org.apache.spark.sql.catalyst.analysis.TableAlreadyExistsException;
+import org.apache.spark.sql.connector.catalog.CatalogPlugin;
+import org.apache.spark.sql.connector.catalog.Identifier;
+import org.apache.spark.sql.connector.catalog.StagedTable;
+import org.apache.spark.sql.connector.catalog.TableCatalog;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.Some;
+import scala.collection.JavaConverters;
+
+/**
+ * Takes a Spark table in the sourceCatalog and attempts to transform it into an Iceberg
+ * Table in the same location with the same identifier. Once complete the identifier which
+ * previously referred to a non-iceberg table will refer to the newly migrated iceberg
+ * table.
+ */
+class Spark3MigrateAction extends Spark3CreateAction {
+  private static final Logger LOG = LoggerFactory.getLogger(Spark3MigrateAction.class);
+  private static final String BACKUP_SUFFIX = "_BACKUP_";
+
+  Spark3MigrateAction(SparkSession spark, CatalogPlugin sourceCatalog, Identifier sourceTableName) {
+    super(spark, sourceCatalog, sourceTableName, sourceCatalog, sourceTableName);
+  }
+
+  @Override
+  public Long execute() {
+    Table icebergTable;
+
+    // Move source table to a new name, halting all modifications and allowing us to stage
+    // the creation of a new Iceberg table in its place
+    String backupName = sourceTableName().name() + BACKUP_SUFFIX;
+    Identifier backupIdentifier = Identifier.of(sourceTableName().namespace(), backupName);
+    try {
+      destCatalog().renameTable(sourceTableName(), backupIdentifier);
+    } catch (NoSuchTableException e) {
+      throw new IllegalArgumentException("Cannot find table to migrate", e);
+    } catch (TableAlreadyExistsException e) {
+      throw new IllegalArgumentException("Cannot rename migration source to backup name", e);
+    }
+
+    StagedTable stagedTable = null;
+    try {
+      Map<String, String> newTableProperties = new ImmutableMap.Builder<String, String>()
+          .put(TableCatalog.PROP_PROVIDER, "iceberg")
+          .putAll(JavaConverters.mapAsJavaMapConverter(v1SourceTable().properties()).asJava())
+          .putAll(tableLocationProperties(sourceTableLocation()))
+          .putAll(additionalProperties())
+          .build();
+
+      stagedTable = destCatalog().stageCreate(Identifier.of(
+          destTableName().namespace(),
+          destTableName().name()),
+          v1SourceTable().schema(),
+          sourcePartitionSpec(),
+          newTableProperties);
+
+      icebergTable = ((SparkTable) stagedTable).table();
+
+      if (!icebergTable.properties().containsKey(TableProperties.DEFAULT_NAME_MAPPING)) {
```

Comment:
Can't this use `stagedTable` properties instead?

Reply:
Sure will change